### ISSUE TITLE
Sub Wars graphical glitches, related to #695

### ISSUE THREAD

**Issue by bunnei** [citra-emu/citra#714]
Sub Wars has some weird glitching that occurs when Citra's software renderer tries to draw the sub in the main title screen. Behavior has varied between platforms (Linux vs. Windows) as well as with and without #695. Initially, we found that the glitching happens only on Windows. #695 makes it show up on Linux (And possibly OS X) We've also found that CiTraces are also affected by the bug in the same manner. @Subv believes that it is a problem with the vertex shader.

### Windows
Without #695: The sub's shadow texture gets rendered first (on top of everything) -> then part of the background scene is rendered (behind the shadow) -> then the sub is rendered with wrong colors -> the rest of the scene is rendered. This causes a green-ish sub with black and white patches, which is wrong. See below:
![](http://i.imgur.com/IhLtEiw.png)
With #695: The sub's shadow texture gets rendered -> part of the background scene gets rendered -> the sub is rendered with wrong colors -> the rest of the scene is rendered -> the sub is completely covered by a white texture? See below:
![](http://i.imgur.com/Xt1H2Cl.png)

### Linux
TBD

**Comment by neobrain** [comment]
The image on Linux shows the ship rendering properly before  #695. After #695, it's glitched like in the screenshots shown for Windows.

**Comment by bunnei** [comment]
@neobrain please add additional clarification: What does "ship rendering properly" look like? Can you post a screenshot? Furthermore, with #695, does it look identical to the first screenshot above (Windows w/o #695) or the second (Windows w/ #695)? If you could just share the screenshots in this issue, that would be helpful. Thanks

**Comment by bunnei** [comment]
So at least for bug part 1, this seems to be an issue with VS vs. GCC. I built Citra on Windows and get:

![](http://i.imgur.com/sey9a4b.png)

**Comment by purpasmart96** [comment]
That is very weird...

**Comment by bunnei** [comment]
Digging some more, the issue appears to be lack of fragment lighting... And more specifically, specular lighting (color combiner source 2, e.g. "SecondaryFragmentColor"). If you change this line -> https://github.com/citra-emu/citra/blob/master/src/video_core/rasterizer.cpp#L410 to `return {0,0,0,0};` on VS, you get the same behavior has Linux:
![](http://i.imgur.com/3e2oIUx.png)
NOTE: These changes are independent of #695. That being said, I'm assigning this issue to @neobrain to implement fragment lighting.

**Comment by bunnei** [comment]
Side note: The text is screwed up still between Linux and Windows. I'm guessing that was also relying on some undefined behavior on GCC.

**Comment by Subv** [comment]
The text should be fixed with my PR, it's probably caused by the hack-fix of the address_offset

**Comment by purpasmart96** [comment]
@bunnei Fragment lighting might be a while... I started doing work on some of it but the regs bit layouts are mostly untested and was based on tfarely's branch, so nothing has been implemented in the software renderer  https://github.com/purpasmart96/citra/tree/tfarley_ogl_hd_add_ons

**Comment by MoochMcGee** [comment]
Note: returning 0 on SecondaryFragmentColor also fixes Dragon Quest VII's warning screen.

**Comment by Subv** [comment]
The name of this PR should be changed to "Fragment Lighting" or something

**Comment by yuriks** [comment]
This is not a PR. :P (Also, I disagree, unless we are 100% certain this == frag lighting.)

**Comment by yuriks** [comment]
This currently looks pretty ok with the OpenGL renderer:
![subwars_hw](https://cloud.githubusercontent.com/assets/341401/9102602/af479cd4-3bca-11e5-820c-b33701c11f96.png)

Software renderer still has the "shiny" coating:
![subwars_sw](https://cloud.githubusercontent.com/assets/341401/9102606/be8060d2-3bca-11e5-9b0c-d33c67f32b1e.png)

(Both look different from hardware, which has a yellow tint matching the scene.)

**Comment by tfarley** [comment]
Forgot to link these: updating to mention that depth offset (#1334) makes the OGL renderer consistent with the software renderer (both get the full shiny coating) and fragment lighting (#1264) gives it the proper yellow tint.

**Comment by tfarley** [comment]
This should be close-able now, unless you want to keep it open until sw renderer has fragment lighting to match.

**Comment by JayFoxRox** [comment]
@citra-emu: as @tfarley statetd: the current bug is not related to #695 anymore so I'd say close this and wait for someone to open a new issue about sw-renderer failing in this game?
